Helpful log msg when NPE in HollowDiffMatcher

nayanika-u opened the pull request.

nayanika-u committed:
Fix NPE in HollowDiffMatcher

Add check for special case of single null parameter

Remove redundant clause inside if condition

nayanika-u commented on hollow/src/main/java/com/netflix/hollow/core/index/HollowPrimaryKeyIndex.java:
```diff
@@ -276,13 +277,18 @@ public int getMatchingOrdinal(Object key1, Object key2, Object key3) {
-        if(fieldPathIndexes.length != keys.length || hashTable.bitsPerElement == 0)
+
+        // for special case of keys containing one null element
+        if (keys == null || (keys != null && fieldPathIndexes.length != keys.length) || hashTable.bitsPerElement == 0)
```
This will throw NPE for cases like getMatchingOrdinal(null). 

getMatchingOrdinal((Object) null), getMatchingOrdinal(null, null), getMatchingOrdinal(new Object[]{null, null}) and so on will pass this check.

Sunjeet replied:
The `keys != null` check is redundant, it can be dropped

nayanika-u replied:
Done.

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/core/index/HollowPrimaryKeyIndex.java:
```diff
@@ -276,13 +277,18 @@ public int getMatchingOrdinal(Object key1, Object key2, Object key3) {
-
+        for (int i = 0;i < keys.length;i++) {
+            // null values can percolate from upstream in the method's Object array
+            if (Objects.nonNull(keys[i])) {
```
Nit: use `!= null` here? Might read better and avoid a tiny performance overhead

Sunjeet replied:
Do we need to add the null check to the non vararg flavors of `getMatchingOrdinal`?

Sunjeet replied:
IIUC a behavior change we're introducing with this change that a component of the primary key is now nullable where earlier it wasn't. I'm not sure what the implication of the computed hashCode would be. Is this something we've already looked into, or do we need to rethink whether we want to support null for key[i] or fail gracefully instead?

nayanika-u replied:
As discussed, this code change might lead to incorrect hashCode computation for cases where it should have just failed (cases with PK passing null field values) and lead to incorrect results in the lookup.

nayanika-u committed:
Logging NPE with relevant information for debugging

Undo changes in HPKI

Undo changes in HPKI

Undo changes in HPKI

Fixing log statement

nayanika-u commented:
Merging this after checking with David.

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/tools/diff/HollowDiffMatcher.java:
```diff
@@ -21,7 +21,10 @@
 import com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadState;
 import com.netflix.hollow.core.util.IntList;
 import com.netflix.hollow.core.util.LongList;
+import com.sun.org.slf4j.internal.Logger;
```
Note: double check import

nayanika-u replied:
changed to java.util.Logging

Sunjeet commented on hollow/src/main/java/com/netflix/hollow/tools/diff/HollowDiffMatcher.java:
```diff
@@ -91,7 +95,14 @@ public void calculateMatches() {
+            } catch(NullPointerException ex) {
+                LOGGER.error("Null values found for type " + toTypeState.getSchema().getName()
+                        + " with key field values " + Arrays.asList(key) + " at ordinal : " + candidateToMatchOrdinal
+                        + " with stacktrace : " + ex.getStackTrace());
```
NOTE: double check that this is logging the stack trace.

nayanika-u replied:
ack. have removed it from printing. As discussed, re throwing this as RuntimeException to preserve existing behavior for Hollow Tooling

akhaku commented on hollow/src/main/java/com/netflix/hollow/tools/diff/HollowDiffMatcher.java:
```diff
@@ -91,7 +95,14 @@ public void calculateMatches() {
         while(candidateToMatchOrdinal != -1) {
             Object key[] = toIdx.getRecordKey(candidateToMatchOrdinal);
-            int matchedOrdinal = fromIdx.getMatchingOrdinal(key);
+            int matchedOrdinal = -1;
```
btw might be good to use `HollowConstants.ORDINAL_NONE` here

nayanika-u replied:
ack. Kept it -1 like other ordinal checks in this class.
